tests/types/resolving/test_optionals.py

```diff
@@ -1,9 +1,65 @@
-from typing import Optional
+from typing import List, Optional, Union
 
 import strawberry
+from strawberry.annotation import StrawberryAnnotation
 from strawberry.type import StrawberryOptional
 
 
+def test_basic_optional():
+    annotation = StrawberryAnnotation(Optional[str])
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type is str
+
+    assert resolved == StrawberryOptional(of_type=str)
+    assert resolved == Optional[str]
+
+
+def test_optional_list():
+    annotation = StrawberryAnnotation(Optional[List[bool]])
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type == List[bool]
+
+    assert resolved == StrawberryOptional(of_type=List[bool])
+    assert resolved == Optional[List[bool]]
+
+
+def test_optional_optional():
+    """Optional[Optional[...]] is squashed by Python to just Optional[...]"""
+    annotation = StrawberryAnnotation(Optional[Optional[bool]])
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type is bool
+
+    assert resolved == StrawberryOptional(of_type=bool)
+    assert resolved == Optional[Optional[bool]]
+    assert resolved == Optional[bool]
+
+
+def test_optional_union():
+    @strawberry.type
+    class CoolType:
+        foo: float
+
+    @strawberry.type
+    class UncoolType:
+        bar: bool
+
+    annotation = StrawberryAnnotation(Optional[Union[CoolType, UncoolType]])
+    resolved = annotation.resolve()
+
+    assert isinstance(resolved, StrawberryOptional)
+    assert resolved.of_type == Union[CoolType, UncoolType]
+
+    assert resolved == StrawberryOptional(of_type=Union[CoolType, UncoolType])
+    assert resolved == Optional[Union[CoolType, UncoolType]]
+
+
+# TODO: move to a field test file
```

Comment:
👀

Reply:
Yeah I didn't want to deal with a complete refactor/reorganize in this PR. Another time